table.cc

```diff
@@ -2446,6 +2449,30 @@ table::run_with_compaction_disabled(std::function<future<> ()> func) {
     });
 }
 
+bool
+table::set_auto_compaction(bool enabled) {
+    if (enabled) {
+        if (_compaction_disabled_by_user) {
+            _compaction_disabled_by_user = false;
+            if (--_compaction_disabled == 0) {
+                // we're turning if on again, use function that does not increment
+                // the counter further.
+                do_trigger_compaction();
+            }
+            return true;
+        }
+    } else {
+        if (!_compaction_disabled_by_user) {
+            ++_compaction_disabled;
+            _compaction_disabled_by_user = true;
+            (void) _compaction_manager.remove(this);
```

Comment:
@glommer yeah, I am reworking that.

Reply:
@glommer, @sitano what if we implement disable_auto_compaction() as 
set_compaction_strategy(sstables::compaction_strategy_type::null) and enable_auto_compaction() as 
set_compaction_strategy(_schema->compaction_strategy());
?

That will take care of backlog as desired given that null strategy uses null backlog tracker.

Reply:
@raphaelsc I thought about this approach, instead of muting backlog directly. The problem with that is that we have Major compaction that still must be allowed. When something forces Major compaction we must turn everything back on including backlog, but as far as we have minor compactions disabled, we must also do not allow Submitting minor (background) compaction pending tasks that will wait for a ReadLock until the Major compaction is happening. 

Thus, we will need to modify Submit() (background compaction submission) behavior to consider all these states anyway. In that case, setting null to the compaction strategy does not look like something that will make the solution simpler.

Reply:
@raphaelsc moreover, there is another thing, user can enforce compactin strategy change from the API. I am not sure it changes schema compaction. In that case we can't mess with null there.

Reply:
@sitano Major compaction would work with null compaction strategy, but as you pointed out, backlog would be an issue. set_compaction_strategy(sstables::compaction_strategy_type::null) wouldn't mess with compaction properties of schema, but only with the in-memory compaction strategy, which is ok. It's only an implementation detail.

Reply:
@raphaelsc btw, looking at the code of major compaction, this approach could work. Major compaction surprisingly uses separately owned additional backlog tracker for handling user-requested compaction:

https://github.com/scylladb/scylla/blob/master/sstables/compaction_manager.cc#L288

```
compaction_backlog_tracker user_initiated(std::make_unique<user_initiated_backlog_tracker>(_compaction_controller.backlog_of_shares(200), _available_memory));
            return do_with(std::move(user_initiated), [this, cf, descriptor = std::move(descriptor)] (compaction_backlog_tracker& bt) mutable {
                register_backlog_tracker(bt);
```

Thus, it could work.

However, what I am missing about picking the compaction strategy by major compaction? I see it uses CF one:

https://github.com/scylladb/scylla/blob/master/sstables/compaction_manager.cc#L278

```
sstables::compaction_strategy cs = cf->get_compaction_strategy();
``` 

So, if we will set it to null, AFAIU it would not work. Did I miss something? In anyway, we could change that also.

Reply:
So, here I misinterpreted what major compaction does with user-initiated additional backlog. But anyway, it seems like everyone loves idea with setting a temporary Null compaction strategy to the table (cf).

Reply:
As far as we have decided not to use NullCompactionStrategy for temporary disabling background compaction due to the unsafe major compaction descriptor evaluation, I am resolving this discussion. (And also because use of ECS is equivalent to the solution with muting backlog IMHO)